### PR TITLE
fix(config): disallow non-unique component names

### LINKED ISSUE #6268: Panic when source and transform have the same name
This should be caught at load time, but it currently results in a panic ([stacktrace](https://gist.github.com/Mitsuwa/098bda4e3218baecdd0a014651fa0770)).

### PR DESCRIPTION
**Pull request by lukesteensen**
Closes #6268

### CODE REVIEW

**Comment by lukesteensen** [comment]
> test sources::host_metrics::tests::filters_disk_metrics_on_device ... FAILED
>  test config::watcher::tests::file_update ... FAILED

Neither of these seems to be related to the change.

File: src/config/validation.rs [hunk 1]
```diff
@@ -7,10 +7,27 @@ pub fn check_shape(config: &Config) -> Result<(), Vec<String>> {
     if config.sources.is_empty() {
         errors.push("No sources defined in the config.".to_owned());
     }
+
     if config.sinks.is_empty() {
         errors.push("No sinks defined in the config.".to_owned());
     }
 
+    // Check for non-unique names across sources, sinks, and transforms
+    let mut name_counts = HashMap::<&str, usize>::new();
```
Comment (bruceg) [inline src/config/validation.rs:16]: Would it be worth trying to tell the user which kind of component was colliding? Then we'd collect into a `HashMap::<&str, Vec<ComponentKind>>`.

File: src/config/validation.rs [hunk 2]
```diff
@@ -7,10 +7,37 @@ pub fn check_shape(config: &Config) -> Result<(), Vec<String>> {
...
+    // Helper for below
+    fn tagged<'a>(
+        tag: &'static str,
+        iter: impl Iterator<Item = &'a String>,
+    ) -> impl Iterator<Item = (String, &'a String)> {
+        iter.map(move |x| (tag.to_owned(), x))
+    }
+
+    // Check for non-unique names across sources, sinks, and transforms
+    let mut name_uses = HashMap::<&str, Vec<String>>::new();
```
Comment (bruceg) [inline src/config/validation.rs:24]: Could this be `HashMap::<&str, Vec<&'static str>>` to avoid cloning the tag?

**Review by bruceg — APPROVED** [review]
Looks good but for a bit of bikeshedding around extra cloning.

File: src/config/validation.rs (hunk 2 above)
Comment (lukesteensen) [reply src/config/validation.rs:24]: Yes it could!

**Pull request merged**